src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java

```diff
@@ -1073,7 +1073,46 @@ private Runnable processTask() {
 				finally {
 					attributesHolder.remove();
 					if (result != null) {
-						this.shardIterator = result.getNextShardIterator();
+                         // If using manual checkpointer, we have to make sure we are allowed to use the next shard iterator
+                         // Because if the manual checkpointer was not set to the latest record, it means there are records to be reprocessed
+                         // and if we use the nextShardIterator, we will be skipping records that need to be reprocessed
+						List<Record> records = result.getRecords();
+						if (CheckpointMode.manual.equals(KinesisMessageDrivenChannelAdapter.this.checkpointMode) &&
+								!records.isEmpty()) {
+							logger.info("Manual checkpointer. Must validate if should use getNextShardIterator()");
+							String lastRecordSequence;
+							if (records.size() == 1) {
+								lastRecordSequence = records.get(0).getSequenceNumber();
+							}
+							else {
+								lastRecordSequence = records.get(records.size() - 1).getSequenceNumber();
+							}
```

Comment:
Doesn't look like this `if..else` gives us some advantages.
Why such a plain `records.get(records.size() - 1)` is not enough to take that last record in the batch?

Reply:
Oh, you're right, I missed that. Fixed it